ui/components/app/snaps/snap-ui-address-input/snap-ui-address-input.tsx

```diff
@@ -0,0 +1,184 @@
+import React, {
+  ChangeEvent,
+  FunctionComponent,
+  useEffect,
+  useRef,
+  useState,
+} from 'react';
+import { useSelector } from 'react-redux';
+import classnames from 'classnames';
+import { debounce } from 'lodash';
+import {
+  CaipAccountId,
+  CaipChainId,
+  parseCaipAccountId,
+} from '@metamask/utils';
+import {
+  Box,
+  FormTextField,
+  FormTextFieldProps,
+  FormTextFieldSize,
+  Icon,
+  IconName,
+  Label,
+  Text,
+} from '../../../component-library';
+import { useSnapInterfaceContext } from '../../../../contexts/snaps';
+import { getAccountInfoByCaipChainId } from '../../../../selectors/selectors';
+import {
+  AlignItems,
+  BackgroundColor,
+  BorderColor,
+  BorderRadius,
+  Display,
+  FlexDirection,
+  FontWeight,
+  IconColor,
+  TextVariant,
+} from '../../../../helpers/constants/design-system';
+import Jazzicon from '../../../ui/jazzicon';
+
+export type SnapUIAddressInputProps = {
+  name: string;
+  form?: string;
+  label?: string;
+  chainId: CaipChainId;
+};
+
+export const SnapUIAddressInput: FunctionComponent<
+  SnapUIAddressInputProps & FormTextFieldProps<'div'>
+> = ({ name, form, label, chainId, error, ...props }) => {
+  const { handleInputChange, getValue, focusedInput, setCurrentFocusedInput } =
+    useSnapInterfaceContext();
+
+  const inputRef = useRef<HTMLDivElement>(null);
+
+  const initialValue = getValue(name, form) as string;
+
+  const [value, setValue] = useState(
+    initialValue
+      ? parseCaipAccountId(initialValue as CaipAccountId).address
+      : '',
+  );
+  const [matchedAddressName, setMatchedAddressName] = useState<string | null>(
+    null,
+  );
+
+  const accountsInfo: Record<string, string> = useSelector((state) =>
+    getAccountInfoByCaipChainId(state, chainId),
+  );
+
+  /*
+   * Focus input if the last focused input was this input
+   * This avoids losing the focus when the UI is re-rendered
+   */
+  useEffect(() => {
+    if (inputRef.current && name === focusedInput) {
+      (inputRef.current.querySelector('input') as HTMLInputElement).focus();
+    }
+  }, [inputRef]);
+
+  const getMatchedAddressName = (address: string) => {
+    const normalizedAddress = address.toLowerCase();
+    return accountsInfo[normalizedAddress];
+  };
+
+  const debouncedHandleInputChange = debounce(
+    (event: ChangeEvent<HTMLInputElement>) => {
+      handleInputChange(name, event.target.value ?? null, form);
+    },
+    80,
+  );
+
+  const handleChange = (event: ChangeEvent<HTMLInputElement>) => {
+    setValue(event.target.value);
+    const matchedName = getMatchedAddressName(event.target.value);
```

Comment:
Can we use https://github.com/MetaMask/metamask-extension/blob/sip-26/ui/hooks/snaps/useDisplayName.ts instead?

Reply:
That way we get both accounts and contacts matched without requiring a new hook